### ISSUE TITLE
1.4.0 - Upload issues

### ISSUE THREAD

**Issue by sunox1** [Nicotine-Plus/nicotine-plus#43]
Hello,

I'm running 1.4.0 on Debian Jessie (stable) 64 bit with a backported kernel (version 4.9).

Uploads seemed to be working just fine after installing 1.4.0, but after a few hours I stopped getting new peers, and I could no longer search my own files from another account. Restarting the program seems to fix the issue, but I don't know how consistent it is. Uploads work fine in SoulseekQT with same listening ports, so it doesn't appear to be a firewall issue.

I should note that I've had similar issues with all versions of nicotine-plus that I have tried so far: the stable version in the main Debian repos, and the main trunk version that was released prior to 1.4.0. 

I'm new to the bug reporting process, so please let me know if I've omitted anything, or there is anything I can clarify. 

Thank you!

**Comment by sunox1** [comment]
sorry, but it looks like there's a good chance my problem was caused by enabling the 'block unresolvable ips' setting. Will repost if that isn't the case. Apologies for the possibly pointless report!

**Comment by sunox1** [comment]
Thought for a second it was a false alarm, but sometime within the last hour uploads have stopped working again.
More notes: I can browse my files and add them to transfers, but their status is "Queued". Most of the time I cannot find my files through the Search function, but occasionally I can for a brief moment or two. When I add the searched-for files to my transfers, their status is "Queued".

**Comment by gfarmerfr** [comment]
Hi @sunox1,

That's quite a strange behaviour indeed.
Are you using UPnP or are you doing a static port redirection ?

**Comment by sunox1** [comment]
Hi @gfarmerfr,

Thanks for the response.

I am forwarding the listening port, and one port directly above it (SoulseekQT requires you to open the listening port and an "obfuscated port", and I do the same with nicotine-plus just to be safe). My router runs OpenWRT, if that matters. I have a local firewall as well, but the appropriate ports are all opened.

**Comment by sunox1** [comment]
Argh, I just got home and it seems to be working for now. When I was having difficulty earlier today I was at school, so maybe there was something about my school's network that was causing problems? Although I was able to download from other users just fine. I'm confused. Anyway, I'll close it again for now. Thanks for your time.

**Comment by gfarmerfr** [comment]
No problem, keeps us updated about the situation.

**Comment by sunox1** [comment]
Hello. Just a little update.

Most of the day I seemed to be sort of working. Although I seemed to getting less uploads than I do with QT (more on that below), I did successfully upload a bunch of files.

As of a couple hours ago, however, my files are unsearchable, and there doesn't seem to be much activity in the debugging window (it seems that when things are working, there are a lot of pierceFirewall and related messages). Restarting the program doesn't fix anything. I can search and download files just fine, however. 

Something I noticed is that around the time when things stopped working, I would get a continual stream of "Can't connect to user [username], giving up" messages. Not sure if that is helpful.

A few final random notes: I run QT and nicotine side by side to help give a baseline for how nicotine is working, and I felt like even when nicotine seemed to be uploading fine, I was getting less uploads than on QT. Also, when things seemed to be working, there were a few instances where I queued up a bunch of files for transfer, but the transfer didn't begin. In one instance, the user trying to download from me got an error message "file not shared", which isn't the first time I've come across that. I googled this error message and in a forum post some commenter said that this meant that there was an issue with port forwarding or file permissions. My permissions are ok (I've even moved all of my files to my home directory), and port forwarding appears to be set up correctly.

**Comment by gfarmerfr** [comment]
Hi @sunox1,

The connection issue at your school doesn't really surprise me because, as you said, you can't do anything about port redirection there. So if you have people in the same situation on the other side (no forwarding) you will see the `giving up` messages. For those who have redirection on the other side it should work.
I'm not entirely familiar how the peers connection scheme works but that's what I'm understanding from this [doc](http://nicotine-plus.org/wiki/SoulseekProtocol#PeerConnectionMessageOrder).
Has QT the same connectivity issues at school by any chance ?

The `I can receive direct connections` checkbox basically mean what it says: it tells Nicotine and other clients that they can contact your PC directly either via a port forwarding (UPnP or static) or if you have a public IP on your PC (bridge mode).
OpenWRT should be tunable enough to activate the bridge mode I think. I know this is not a desirable state but have you tried it to see how Nicotine would react ? An alternative to test, if OpenWRT permits it, is to put your PC in a DMZ (should be a similar behaviour that the bridge mode).

**Comment by sunox1** [comment]
Hi @gfarmerfr,

Thanks for the response, and I appreciate the help. I will give that a try tonight.

Sorry, I wasn't clear about the school thing. I was on my Windows laptop using QT and trying to download from my home pc running nicotine. I tried downloading from a bunch of different people, and my home pc was the only account that gave me problems. Today I did the same thing though and I can download from my home pc running nicotine just fine, so I'm no longer sure what to make of what happened the other day. 

The `can't connect...` errors and all the other errors appear on the logs of my home machine when idling (and also when I check the `I can receive direct connections` box.

The behaviour of nicotine after I check the `I can receive...` box definitely means that port forwarding is the issue then. I suppose it is now a question of whether it is my network config or the client that is causing the issue. It doesn't seem like my network config is the issue, because both QT and nicotine report that the ports are being forwarded, QT appears to be working perfectly fine with in the same environment, and in the past I have successfully port forwarded with other applications.

I'm not certain that I am doing port forwarding corectly, but this is how I am doing it at home (I arbitrarily chose a 5 digit listening port - let's say it's 20000): myRouter : 20000 => myPC (static IP) : 20000. I suppose there could be something wrong with that, but it would be strange considering that a number of indicators all say that it is being done correctly. It also wouldn't explain why QT seems to work perfectly fine with the exact same client settings and network config: I upload approximately twice as many files on QT as on nicotine-plus on the same machine sharing the same files (I just listen on different ports). 

I will try playing with the router settings tonight when I get home later, and hopefully I will find something.

**Comment by sunox1** [comment]
Hi @gfarmerfr,

I was able to put my pc into a dmz, and nicotine behaves the same way: If I check `I can receive direct connections` I am unable to make any connections whatsoever, and if I leave this box unchecked it seems I can make connections with some but not all.

And to make extra sure my ports are properly being forwarded, I checked them with a few port checker websites such as http://portchecker.co/ and http://www.canyouseeme.org/, and they all show it as being open.

**Comment by sunox1** [comment]
I've also managed to shrink down my former novel on the issues (a few posts up) to just a few paragraphs. I've learned that a lot of what I was discussing as potential problems are actually expected behaviour. I've pasted it below for visibility's sake.

**Uploading problems**
When the I can receive direct connections box is unchecked, I get about half the uploads as I do in QT, which I assume is expected behaviour. Occasionally, I become unreachable via search, and I can't form connections with anyone at all, which is a mystery to me.

When the I can receive direct connections box is checked, however, it's a different story. It oscillates back and forth between seeming to be working fine (haven't been able to compare it with QT yet, so can't confirm), or being totally unreachable (can't find my files via search, and lots of Can't connect... messages in the log).

**Port status**
This seems to be the big clue: When I am in the unreachable state, the nicotine port checker shows the ports as closed, and open otherwise. A third party port checker website corroborates what the nicotine checker says. The QT port checker, by contrast, always reports as being open.

**I think the ports are being forwarded correctly**
I don't know exactly how to interpret this, but it doesn't seem possible that my router would be responsible for this, since QT's ports aren't being opened and closed, QT is working fine in general, and same with the other applications for which I have forwarded ports for in the past.

Thanks again.

**Comment by gfarmerfr** [comment]
Hi @sunox1,

Great debugging :)
You're right : the fact that N+ behaviour doesn't change in a DMZ clearly indicate that it's not a port forwarding issue.

Just to be sure there's no config corruption can you provide the two following values from your `~/.nicotine/config` file:
```
firewalled = XXX
portrange = (YYY, ZZZ)
```

As you previously indicate you have a local firewall: I suppose your firewall is configured to have the ports in the portrange open, yes ?

The fact that the port status oscillates is quite puzzling.
A few wild guess that might be worth exploring when you observe the "port closed" status:

1) The router for some reason has stop to forward traffic to N+ port (very unlikely): no simple way to test that, maybe combining ```nc``` and ```tcpdump``` ?
2) N+ local socket is dead: can be checked via this command line (replace 2234 by the port you've configured N+ to listen on, most likely it will be the first port of your portrange):
```
netstat -laputen | grep 2234   
tcp        0      0 0.0.0.0:2234            0.0.0.0:*               LISTEN      1000       11648341   11838/python
...
```
3) The number of socket connections has exceed the limit (maybe a leak in the connection pool ?)
You can check your limit via this command (replace 11838 with n+ pid):
```
cat /proc/11838/limits | grep -i "Max open files"
Max open files            4096                4096                 files    
```
and the current number of open socket by N+ is in the bottom right status bar.
N+ won't go further that 90% of the "Max open files".

4) Sometime I've seen SYN flood reported by the kernel via ```dmesg```
```
[660395.150814] TCP: request_sock_TCP: Possible SYN flooding on port 2234. Sending cookies.  Check SNMP counters.
```
I'm not sure of the consequences of this and how your firewall is configured to react to that particular nasty behaviour.

If everything is correct for these parts we can definitively be sure that N+ has some bug lurking around that I haven't identified yet.

**Comment by sunox1** [comment]
Thank you for the advice. I haven't got a definitive answer for 1. yet, but I was able to get answers for everything else.

In ~/.nicotine/config:
`  firewalled = False
  portrange = 55525, 55531`

I do normally a firewall enabled and all of those ports open, but I disabled it last night just so I don't have to worry about it as a possible cause of any issues.

1. I looked at the output of `tcpdump | grep 55525` during the transition from `open` to `close` and I didn't see anything out of the ordinary. Looking at tcpdump, I don't think you can tell anything is wrong. I can keep trying to look. 

I am reading about `nc` now.

If the router stop forwarding ports and then starts again, would this be logged in my router's logfile? I checked there isn't anything there.

2. Everything looks good in `netstat`: my first line looks identical to your sample output.

3. My `Max open files` is 2048, which is higher than N+'s self-imposed limit of 1843.

4. It looks like I get the same message a few times per day, except instead of `Sending cookies.` mine says `Dropping request.`. These messages happen far less frequently than port oscillations, so maybe that means they aren't related? Also, not sure if this is relevant or not, but last night I actually turned off SYN flood protection in OpenWRT to see if it would help anything.

Another thing that seems significant is that it looks like the N+ listening port is always open when I start nicotine, and stays open for a minute or two. It seems like it has to 'build up' to the problem state.

**Comment by gfarmerfr** [comment]
Re @sunox1,

1) I'm not sure what the router will log.
2) OK
3) OK
4) I've research further and these kind of messages should only be triggered when lots of connections are established in a short amount of time, like when you do a search. So it should not be related to your problem from what I understand.

Researching 4) I've put together a experimental patch in one of my repo branch https://github.com/gfarmerfr/nicotine-plus/tree/slskproto-put-nowait
I doubt it will change anything to your problem but it wouldn't hurt to try I guess...

**Comment by sunox1** [comment]
I think I switched branches successfully. I cloned the main repo and then did `git checkout -b /origin/slskproto-put-nowait`, which seemed to work, and then I launched `nicotine` from the directory root. The version number was still 1.4.0, though.

Anyway, no change unfortunately.

I mentioned above that the ports close less often if I share fewer files. I found later that if I uncheck `Enable sending search results to other users`, the port never closes. If I enable search and set it to a maximum of one, it seems to close just as often as it does at 50. I'm not sure if this is interesting or useful at all, but just thought I would mention.

**Comment by sunox1** [comment]
So it turns out I wasn't installing the branch correctly. I finally figured out how to do so

AND

I've been seeding all my files for about 15 minutes now without any port closures!!!! Whether it is totally fixed, I don't know, but that is at the very least a massive improvement.

I'm just heading out for the day, but I will let it run alongside QT and we will see how it goes. I hesitate to say it is fixed, but things look very promising right now ! I will provide an update by tonight (EST) at the latest.

**Comment by sunox1** [comment]
So the issue is indeed solved (woo!), and it looks like the reason is that I upgraded to Debian Stretch. I could have swore that I tested n+ after upgrading and still encountered the port closing issue, but maybe not :/

I *think* that takes care of all the issues for now, so I'll to close this. Thanks so much for your time and effort - I really appreciate it. I learned a fair amount during this process, and I hope it wasn't a total waste of your time.

Cheers!

**Comment by gfarmerfr** [comment]
No problem.
So the problem has disappear by upgrading debian alone or did you still need to run the experimental patch on top of the upgrade ?
I'm a little lost and maybe tired too :)

**Comment by sunox1** [comment]
Sorry, that wasn't clear.

It looks like just upgrading from Debian Jessie to Stretch (stable to testing) fixed it. Late yesterday I downgraded to the N+ version in the Debian testing repos (quite old) and everything ran fine - no port closures. I originally thought the patch had fixed because I thought (mistakenly) I tested the master version in Stretch, before trying the experimental patch, and observed the same old bug behaviour. I guess I started to imagine things after staring at logs for 2 days! :)

Luckily, Stretch will become the Stable release in a few months, so this should be a non-issue going forward.
